tasmota/berry/drivers/Core200S.be

```diff
@@ -0,0 +1,712 @@
+    end
+
+    # read an remove from buffer
+    def consume(num)
+        var content = self.read_ahead(num)
+        if content != nil
+            self.read_buffer = self.read_buffer[num..]
+        end
+        return content
+    end
+
+    def read_msg(allow_no_msg)
+        var max_read_attempts = 10
+        var header = self.read_ahead(6)
+        if header == nil
+            header = bytes("0000")
+        end
+        while header[0] != 0xa5 || (header[1] != 0x12 && header[1] != 0x22)
+            if header != bytes("0000")
+                self.consume(1)
+            else
+                max_read_attempts -= 1
+                if max_read_attempts <= 0
+                    if !allow_no_msg
+                        info_c2s("Too many read attempts (couldn't read message header)")
+                        self.error_count += 1
+                    end
+                    return bytes("")
+                end
+                tasmota.delay(100)
+            end
+            header = self.read_ahead(6)
+            if header == nil
+                header = bytes("0000")
+            end
+        end
+
+        if header == bytes("0000")
+            info_c2s("Too many read attempts (couldn't read message header)")
+            self.error_count += 1
+            return bytes("")
+        else
+            self.consume(6)
+        end
+
+        log_c2s("Got Header " + header.tohex())
+
+        var payload = self.consume(header[3])
+        while payload == nil
+            max_read_attempts -= 1
+            if max_read_attempts <= 0
+                if !allow_no_msg
+                    self.error_count += 1
+                    info_c2s("Too many read attempts (couldn't read message payload)")
+                end
+                return bytes("")
+            end
+            payload = self.consume(header[3])
+        end
+
+        var msg = header + payload
+        info_c2s("Full Message " + msg.tohex())
+
+        self.expected_replies -= 1
+
+        var checksum = self.calc_checksum(msg)
+        if checksum != msg[5]
+            info_c2s("Invalid Checksum")
+            return bytes("")
+        end
+
+        return msg
+
+    end
+
+    def check_errors()
+        return self.error_count > 10 || self.expected_replies > 10
+    end
+
+    def reset_counters()
+        self.error_count = 0
+        self.expected_replies = 0
+        self.msg_count = 0
+    end
+end
+
+ser = Core200SSerial()
+
+class Core200S
+    var wifi_on
+    var second_counter
+    var fan_mode
+    var sl_mode
+    var fan_speed_mode
+    var dp_mode
+    var dp_auto_off_mode
+    var cl_mode
+    var nl_mode
+    var timer_remaining
+    var timer_total
+
+    def init()
+        self.wifi_on = nil
+        self.second_counter = 0
+        self.timer_remaining = 0
+        self.timer_total = 0
+        self.init_cmds()
+    end
+
+    def init_cmds()
+        ser.flush()
+        ser.reset_counters()
+
+        tasmota.delay(1000)
```

Comment:
Use async function or state machine. You can't stop Tasmota for 1 full second.

Reply:
hmm, not sure if async behavior is easy to implement here (because the messages kind of need to be guaranteed to be send in this order, or at least first ones). If've seen weird behavior if the order of messages changes. So maybe some simple state machine needs to do the trick.

Reply:
See above